types/mdx/types.d.ts

```diff
@@ -57,13 +57,43 @@ interface NestedMDXComponents {
 
 // Public MDX helper types
 
+/**
+ * You can augment this interface to define custom [components](https://mdxjs.com/guides/injecting-components/).
+ *
+ * This provides type safety when using the `components` prop or a components provider.
+ *
+ * @example
+ * ```ts
+ * import * as React from 'react'
+ *
+ * function Button(props: React.ComponentPropsWithoutRef<'button'>) {
+ *   return <button {...props} />
+ * }
+ *
+ * function Input(props: React.ComponentPropsWithoutRef<'input'>) {
+ *   return <input {...props} />
+ * }
+ *
+ * declare module 'mdx/types.js' {
+ *   interface CustomComponents {
+ *     Button: typeof Button
+ *     Namespace: {
+ *       Input: typeof Input
+ *     }
+ *   }
+ * }
+ * ```
+ */
+export interface CustomComponents {}
```

Comment:
- Not sure about the name. They’re not custom.
- the empty interface by default isn’t an empty interface, perhaps there needs to be a `unique symbol: never` (something like that) in there.
- why after `Nested` but before `Key in StringComponent`?
- “This provides type safety when using the `components` prop or a components provider.” Does it, on its own? I don’t think the last 5 lines here are more type safe than without:

    ```tsx
    function Planet({name}) {
      return <div>Hello {name}</div>
    }

    <Content components={{Planet}}>

    declare module 'mdx/types.js' {
      interface CustomComponents {
        Planet: typeof Planet
      }
    }
    ```

Reply:
- do you want folks to type their `h1`s and such here too? Or just the “capitalized” ones
- What if someone puts `Planet` in an MDX file? So, locally defined components, should they also be augmented?

Reply:
> * Not sure about the name. They’re not custom.

I don’t follow, but I’m open to an alternative name.

> * the empty interface by default isn’t an empty interface, perhaps there needs to be a `unique symbol: never` (something like that) in there.

I don’t understand. Are you referring that interfaces are _shapes_, not _types_? That’s not relevant here.

> * why after `Nested` but before `Key in StringComponent`?

No reason. It had to go somewhere.

> * “This provides type safety when using the `components` prop or a components provider.” Does it, on its own? I don’t think the last 5 lines here are more type safe than without:
>   ```tsx
>   function Planet({name}) {
>     return <div>Hello {name}</div>
>   }
>   
>   <Content components={{Planet}}>
>   
>   declare module 'mdx/types.js' {
>     interface CustomComponents {
>       Planet: typeof Planet
>     }
>   }
>   ```

Assuming `Planet` has type definitions

```tsx
function Planet({name}: {name: string}) {
  return <div>Hello {name}</div>
}
```

This is the type safe line now.

```tsx
<Content components={{Planet}}>
```

These are now errors.

```tsx
<Content components={{
  // error: Planet is missing
}}>

<Content components={{
  // error: Galaxy is not assignable to Planet.
  Planet: Galaxy
}}>
```

But more importantly, this can now be checked by `@mdx-js/language-server`:

```mdx
# Solar system

<Planet
  {/* error: name must be string, not number */}
  name={42}
/>

{/* This is still fine */}
<Asteroid />

export function Asteroid() {
  return <div />
}
```

This can also be supported in `@mdx-js/language-server` without involving `@types/mdx`, but I like the idea that they work together.

Reply:
> > the empty interface by default isn’t an empty interface, perhaps there needs to be a unique symbol: never (something like that) in there.
>
> I don’t understand. Are you referring that interfaces are shapes, not types? That’s not relevant here.

An empty interface in TypeScript does very little: any non-nullable value is assignable to `{}`. [(reference)](https://typescript-eslint.io/rules/no-empty-interface/)
instead consider https://github.com/sindresorhus/type-fest/blob/main/source/empty-object.d.ts

Reply:
An empty interface is `object`, meaning that folks can assign whatever they want.
This bug currently exists in unist/hast/mdast/etc, where `interface Data {}` is used for `node.data`. Anyone can assign anything they want there.
When the first person *does* register something, it becomes an actual interface and types are suddenly checked.

As Christian mentions: https://github.com/unifiedjs/unified/blob/b69689bba52a918d87aa62f295ccffa8d9aa8ef8/index.d.ts#L25-L26

> I don’t follow, but I’m open to an alternative name.

If you write some JSX with say React, say: `<Something><something /><some.thing /><some:thing /></Something>`. None of these things are called “custom component”. That’s not a word currently in use for JSX or MDX as far as I am aware.

The difference in JSX is more “reference” (variable/member expression) vs “literal” (string). We already support that. So it isn’t that either. You can define things and reference them.

This PR is more about how TS looks things up? Either references to the local scope, literals in `JSX.IntrinsicElements`, or in this new registry?
Maybe we could also do an `MDX.Components` registry? `MDX.IntrinsicComponents`

Reply:
And what happens if a local `Planet` is defined, compared to this type?
Or, when someone defines `div`, which conflicts with `JSX.IntrinsicElements`?